### ISSUE TITLE
3.x/4.x Officially support MariaDB

### ISSUE THREAD

**Issue by inoas** [cakephp/cakephp#11340]
This is a (multiple allowed):

* [x] enhancement
* [x] feature-discussion (RFC)

* CakePHP Version: 3.next / 3.5
* Platform and Target: MariaDB 10.2

### What you did
Using MariaDB since ever, even in CakePHP 2 (Galera Cluster).

### What happened
- The docs do not really mention MariaDB consistently
- While it used to be API compatible and mostly is (so I saw different SQL Errors Messages for the same error, just this week), shouldn't we start to support it officially?
- Can we add test configurations for MariaDB to Travis?

### What you expected to happen
MariaDB to be supported:
1. We can support MariaDB 5.5+ as it is API compatible with MySQL 5.5.3+ anyway.
2. We can also opt to forgo those old versions and directly support MariaDB 10.2+ **officially only**, doesn't mean it won't run on lower versions but then we have the option in 4.x to create a different MariaDB driver which then supports fancy features like [Common Table Expressions](https://github.com/cakephp/cakephp/issues/10666) (MariaDB 10.2.1+) and [Window Functions](https://github.com/cakephp/cakephp/issues/10670) (10.2.0) or [Constraint Checks](https://mariadb.com/kb/en/library/constraint/) (10.2.0), even if MySQL does not (yet).
3. In future people could choose to run the MySQL driver which would work on 5.5.3+ (or in 4.x a higher but still old version) with their older MariaDB anyway. So that could mean we could be progressive about a MariaDB driver targeting 10.2.+ during CakePHP 4.x lifetime?!

Ref: https://mariadb.org/feedback_plugin/stats/major_version_breakdown/

**Comment by markstory** [comment]
We don't specifically test/support/mention percona-server. Why is MariaDB different than percona or other MySQL like databases that requires explicit mention/testing?

**Comment by inoas** [comment]
### TLDR

- Percona is a performance / reliability / scalability fork which little features added. No reason to create a new driver here; it lumps behind MySQL / stays on par with MySQL (I can be wrong).
- MariaDB on the other hand is a fork that implements features years ahead of MySQL.

### Explanation

Well I don't know Percona pretty much but as far as I know Percona is a lot closer to MySQL. It is basically a fork that focuses on performance, scalability, reliability. It is closely related to Oracle MySQL still. For instance https://www.percona.com/live/17/sites/default/files/slides/CTE-PLAms2016_0.pdf show that they expect CTE to be integrated into Percona *too* once MySQL 8.0 hits.

On the other hand MariaDB is much more like LibreOffice. The early fork was pretty much the same (MariaDB 5.5) but MariaDB 10.2 has a lot of features that MySQL doesn't have.

For instance MariaDB has complex constraint checks which MySQL still doesn't have: https://bugs.mysql.com/bug.php?id=3464

This is partially due to MariaDB being driven by many of of the previous Sun's core team around the original MySQL author Monty / Michael Widenius.

**Comment by ADmad** [comment]
@inoas You can create a plugin with mariadb specific driver. Once it's popular enough and/or enough people call for it's inclusion in core,  we can add it to the core.

**Comment by inoas** [comment]
A dedicated driver would only make sense if it supported features like CTEs and Window Functions. Not sure if a driver-change alone would be sufficient - also then Postgres and MS Sql Server should receive the same upgrade as they also support CTEs and Window Functions; just MySQL < 8.0 Alpha (not even released) does not.

**Comment by inoas** [comment]
Not sure I will find the time, but if I do, would it be okay to only support MariaDB 10.2+.
For anything before peeps can use our MySQL driver (5.1 / 5.5 compatible) with MariaDB (I have been for more than 2 years without problems).

**Comment by davidyell** [comment]
We also run our Cake apps on MariaDB, but are limited to the version we can use to one which is MySQL compatible.

**Comment by markstory** [comment]
If a separate driver class is needed, I could see an argument for having explicit testing & mention. With the same driver I don't see the point.

If MariaDB does need a separate driver implementation to get connectivity with new versions, that is something we could do, but it will need interest and time to get off the ground.

**Comment by ravage84** [comment]
Totally agree with @markstory. Closing then.
@ionas  if you see some addition to be made to our documentation, we can do that.
